### PR TITLE
Warning for slashes in domains

### PR DESCRIPTION
**Pull request by whatyouhide**
Implements the ideas discussed in #75.

### CODE REVIEW

**Comment by josevalim** [comment]
If the domain has a / in it, we can say in the error message that slashes in domain names are not supported. For this case, we may not even need to show the available domains. I don't remember why we would want to show the known domains anymore, hahaha (sorry!).

File: lib/gettext/compiler.ex [hunk 1]
```diff
@@ -167,13 +174,34 @@ defmodule Gettext.Compiler do
...
+  `backend` is used to print a more helpful warning by adding a list of the
+  domains known by the backend.
+
+  This function is called by `lgettext` and `lngettext`.
+  """
+  @spec warn_if_domain_has_slashes(module, binary) :: :ok
+  def warn_if_domain_has_slashes(backend, domain) do
+    if String.contains?(domain, "/") do
+      IO.puts :stderr, """
+      The domain #{inspect domain} was not compiled. The #{inspect backend}
```
Comment (josevalim) [inline lib/gettext/compiler.ex]: warning: the domain ...

**Comment by whatyouhide** [comment]
@josevalim makes sense :) I squashed the second commit, now we warn when the domain is unknown (and we show the available domains) and we warn that `/`s in domains aren't supported if the domain contains any `/`s.

File: lib/gettext/compiler.ex [hunk 2]
```diff
@@ -167,13 +174,34 @@ defmodule Gettext.Compiler do
...
+
+  This function is called by `lgettext` and `lngettext`.
+  """
+  @spec warn_unknown_domain(module, binary) :: :ok
+  def warn_unknown_domain(backend, domain) do
+    known = backend.__gettext__(:known_domains)
+    cond do
+      String.contains?(domain, "/") ->
+        IO.puts :stderr, "warning: slashes in domains are not supported: #{inspect domain}"
+      not domain in known ->
```
Comment (josevalim) [inline lib/gettext/compiler.ex]: Discussion point: should we really warn here? Should we ask folks to create empty .po files in their locales then? I would say we shouldn't warn here (we can still keep known_domains around, I would just remove the warning. If we warn, we will break the workflow where I can just add translations, specially in english, without caring about the .po / .pot files.

**Comment by josevalim** [comment]
Sorry I was not super clear in my previous comment(s).  My point is simply that we can keep known_domains around but I don't think we would use it for anything right now. I think the warning will cause more confusion than help. :)

**Comment by josevalim** [comment]
And to be super clear: the warning on slashes is still super useful.

File: lib/gettext/compiler.ex (hunk 2 above)
Comment (lexmag) [inline lib/gettext/compiler.ex]: I'm :+1: for not having this warning or make it configurable, e.g. `allow_missing: false`. :bowtie:

File: lib/gettext/compiler.ex [hunk 3]
```diff
@@ -167,13 +174,34 @@ defmodule Gettext.Compiler do
   end
 
   @doc """
+  Prints a warning on `:stderr` if `domain` isn't known to `backend`.
+
+  `backend` is used to print a more helpful warning by adding a list of the
+  domains known by the backend.
+
+  This function is called by `lgettext` and `lngettext`.
+  """
```
Comment (lexmag) [inline lib/gettext/compiler.ex]: Maybe it should be `@doc false`?

Comment (whatyouhide) [inline lib/gettext/compiler.ex]: The whole module is `@moduledoc false`, all other functions are documented with `@doc`.

File: lib/gettext/compiler.ex (hunk 2 above)
Comment (whatyouhide) [inline lib/gettext/compiler.ex]: Honestly, I don't want to warn on missing domains. It's a little bit of a mess, we have to capture so much IO in the tests and @josevalim makes a great point (you'll be warned **everytime** until you run `mix gettext.extract`).

**Comment by whatyouhide** [comment]
Ok, we're superclear now: we'll keep the known locales around (cause it's easy and it can be useful) and we'll only warn for domains with `/`s in them.

**Comment by lexmag** [comment]
> ... we'll keep the known locales around (cause it's easy and it can be useful) ...

You mean known domains, right?
If so, why we keep them if there is no use for it – dead code.

**Comment by josevalim** [comment]
Feel free to remove known_domains. but it doesn't hurt to keep it as well.

**Comment by whatyouhide** [comment]
Ok @josevalim and @lexmag, I decided to remove `known_domains`-related stuff as it's indeed dead code and it's something internal to Gettext, so it wouldn't benefit users; we can always re-add it later (as it's terribly simple to re-implement).

I also fixed the warnings: now only domains with slashes in them emit warnings when used. Ready to merge?

**Comment by josevalim** [comment]
:shipit:

**Comment by vborodkin** [comment]
Hi there! I might be missing something, but how do you treat nested domains then? For example, `dgettext "users/admin", "List Users"`, where the domain file should inside a `users` directory.

**Comment by whatyouhide** [comment]
@vborodkin is this something GNU gettext supports?

**Comment by vborodkin** [comment]
@whatyouhide I don't know. I never used gettext until I started using Phoenix, which is recently. Is there anything I can do to help and clarify this? Although, it seems fairly normal to namespace domains in larger applications. I can use a underscore for the time being.

**Comment by OvermindDL1** [comment]
GNU gettext does support `/` in domains just for note (I just tested it).

**Pull request merged**